pkg/dns/dns_test.go

```diff
@@ -0,0 +1,282 @@
+package dns
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/proto-public/pbdns"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+)
+
+type MockedNetConn struct {
+	net.Conn
+	mock.Mock
+}
+
+type DNSTestSuite struct {
+	suite.Suite
+}
+
+// func (s *DNSTestSuite) SetupTest() {
+
+// }
+
+// func (s *DNSTestSuite) AfterTest() {
+
+// }
+
+func TestDNS_suite(t *testing.T) {
+	suite.Run(t, new(DNSTestSuite))
+}
+
+func genRandomBytes(size int) (blk []byte) {
+	blk = make([]byte, size)
+	_, _ = rand.Read(blk)
+	return blk
+}
+
+func (s *DNSTestSuite) Test_DisabledServer() {
+	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
+	server, err := NewDNSServer(DNSServerParams{
+		BindAddr: "127.0.0.1",
+		Port:     -1, // disabled server
+		Logger:   hclog.Default(),
+		Client:   mockedDNSConsulClient,
+	})
+	if err != nil {
+		s.T().FailNow()
+	}
+	err = server.Run()
+	s.Require().Equal(ErrServerDisabled, err)
+	s.Require().Equal(server.TcpPort(), -1)
+	s.Require().Equal(server.UdpPort(), -1)
+	server.Stop()
+
+}
+
+func (s *DNSTestSuite) Test_ServerStop() {
+	mockedDNSConsulClient := pbdns.NewMockDNSServiceClient(s.T())
+	server, err := NewDNSServer(DNSServerParams{
+		BindAddr: "127.0.0.1",
+		Port:     0, // let the os choose a port
+		Logger:   hclog.Default(),
+		Client:   mockedDNSConsulClient,
+	})
+	if err != nil {
+		s.T().FailNow()
+	}
+	err = server.Run()
+	if err != nil {
+		s.T().FailNow()
+	}
+	server.Stop()
+
+	s.Require().Eventually(func() bool {
+		port := server.TcpPort()
```

Comment:
I would probably grab the port before the call to `server.Stop()` As is this seems to be relying on the fact that the state of the listener will not change when the connection is closed which seems like an implementation detail we may not want to rely on.